### ISSUE TITLE
Rename `exception.*` to `error.*` (aligning with ECS)

### ISSUE THREAD

**Issue by trask** [open-telemetry/opentelemetry-specification#3198]
**What are you trying to achieve?**

Rename `exception.*` attributes to the corresponding ECS names under `error.*` (see https://github.com/open-telemetry/oteps/pull/222 for motivation).

Rename:

* `exception.type` -> [`error.type`](https://www.elastic.co/guide/en/ecs/current/ecs-error.html#field-error-type)
* `exception.message` -> [`error.message`](https://www.elastic.co/guide/en/ecs/current/ecs-error.html#field-error-message)
* `exception.stacktrace` -> [`error.stack_trace`](https://www.elastic.co/guide/en/ecs/current/ecs-error.html#field-error-stack-trace)
* `exception.escaped` -> (nothing corresponding)

**Comment by trask** [comment]
@AlexanderWert it's not clear to us the best mapping from OTel's all-in-one-string (unstructured) `exception.stacktrace` to ECS's multi-field/wildcard `error.stack_trace`.

Is `error.stack_trace.text` the convention for capturing an all-in-one-string (unstructured) stack trace?

**Comment by AlexanderWert** [comment]
@trask In ECS (and Elasticsearch) there are different types of strings: `keyword` (indexed as is for direct access) and `text` (optimized for search in text). A field can have both characteristics, in that case `fieldname` is the keyword and `fieldname.text` is the `text` type field.

`error.stack_trace` is such a `wildcard` type which is a special case of `keyword` type. And it also searchable, so it has in addition the `error.stack_trace.text` field. 

For OTel semconv the `keyword` / `wildcard` typed field is relevant and we can ignore the `text` field.

**Comment by mwear** [comment]
I don't have strong feelings on this either way, but wanted to link two previous issues where this was discussed. #427 and its successor #697. The original proposal named the conventions `error.*`. It was changed due to the discussions on those PRs.

**Comment by trask** [comment]
Some potential reasons not to rename `exception.*` to `error.*`:

* the (stable) API defines the operation as [`RecordException`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#record-exception)
* this would be a breaking change to the telemetry emitted by the (stable) API
* the reasoning at the time for choosing "exception" over "error" was [we record an exception, which may or may not be an error](https://github.com/open-telemetry/opentelemetry-specification/pull/697#issuecomment-656742421)

**Comment by trask** [comment]
> Thinking more about this

same...

I'd like to see if we can avoid having both `error.*` and `exception.*`, as that seems a bit confusing.

And I think `error.*` is the more general term, so I'd lean towards keeping that (it seems reasonable to say exceptions are a kind of error).

I think the existing API method `Span#RecordException` is still accurate, since it specifically records an exception on a span. The fact that it would record it using (the more general) attributes `error.*` seems reasonable.

**Comment by Oberon00** [comment]
> And I think error.* is the more general term, so I'd lean towards keeping that (it seems reasonable to say exceptions are a kind of error).

Hmm, but not in the sense of the original argument of renaming. An exception may indicate that some problem occurred but was handled, in which case, was it really an error? Or an exception might be used for other things, like Python's [GeneratorExit](https://docs.python.org/3/library/exceptions.html#GeneratorExit),

**Comment by trask** [comment]
ya, I don't disagree, I guess my preference is to stretch the meaning a bit instead of having both `error.*` and `exception.*` since the distinction is pretty subtle for most(?) users

**Comment by pyohannes** [comment]
> I'd like to see if we can avoid having both `error.*` and `exception.*`, as that seems a bit confusing.

I agree. Having both `error` and `exception` namespaces with `type`, `message`, and `stacktrace` fields would be very confusing.

> And I think `error.*` is the more general term, so I'd lean towards keeping that (it seems reasonable to say exceptions are a kind of error).

I agree with the first part of the statement. An error is a generic concept, whereas an exception is a language specific construct. Not all languages have an exception construct, e. g. [Go](https://github.com/open-telemetry/opentelemetry-go/blob/main/trace/trace.go#L360) and [Rust](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-api/src/trace/span.rs#L72).

To summarize, I'd be strongly in favor of having just one of `error` or `exception`. I have no strong opinion on which one of these two to settle, but I have slight preference for `error` for the following reasons:
* An "error" is a more generic concept than an "exception". There is value in capturing errors that are no exceptions.
* For telemetry consumers, the crucial distinction often is not whether something is an error or an exception, but whether something is expected or unexpected. Both errors and exceptions can be either expected or unexpected and need to be interpreted in context. For HTTP client spans with 4xx status codes we require the span kind to be set to `ERROR`, but this can be expected (e. g. for checking whether a certain DB entry does _not_ exist). With that in mind, I don't see much benefit in using the "exception" term rather than "error".

**Comment by Oberon00** [comment]
What about the Span status? An error can be recorded by setting the span status to ERROR and setting a appropriate status text. Cf. #306.

I think there is totally space for separate error and exception conventions, and they might not even look that similar: https://github.com/open-telemetry/opentelemetry-specification/issues/306#issuecomment-542579242

> Actually, that ECONNREFUSED highlights another use case that is not well-supported by that statusCode field: If we had some generic convention for OS errors, you could do something like setting os.errcode = "ECONNREFUSED", os.errfunc = "connect" on an HTTP span and you would know that there was an OS-level error, not an HTTP-level error.
>
> EDIT: Or maybe err.kind = "os", err.code = "ECONNREFUSED", err.os.func = "connect". Could be adapted to err.kind = "exception" err.code = "java.net.ConnectException", err.msg = "Connection to [::2]:14426 refused.". We need semantic conventions for errors and exceptions!

Since you mention Rust and Go, they both have both panics and error/Result types, and rust also has a [ControlFlow](https://doc.rust-lang.org/stable/std/ops/enum.ControlFlow.html) mechanism that is a generalization of the "Result" "error" type.

More on errors vs exceptions from @jmacd at https://github.com/open-telemetry/opentelemetry-specification/issues/306#issuecomment-630940573:

> On the other hand, I support the notion that the existing status codes are sufficiently and usefully generic. In the SIG call, one example was given: suppose the user has a "language exception". How should this map into a status code? Language exceptions can mean anything: some language exceptions will map to deadline exceeded, some to permission denied, some to not found, etc. It's fine to leave the default, which is "OK", which gives less information to the backend, but is "correct".

Here, also from @jmacd on the original name decision: https://github.com/open-telemetry/opentelemetry-specification/pull/427#issuecomment-642970559

> I would support changing "error.type" and "error.message" to "exception.type" and "exception.message". It's OK to have an exception that is not an error. When you wrote throw new RuntimeException("message") that translates into exception.type=java.lang.RuntimeException and exception.message=message.
>
> Likewise, it's fine to record a stacktrace for a non-error. Can we just call the stacktrace attrribute stacktrace? That's unambiguous, I think.
>[...]


=> So the intention of the OTel exception.* conventions is to be used _just_ for exceptions. For errors (and even to indicate whether an exception is an error or not), you'd use other conventions (e.g. http.status_code) or the span Status (message+ERROR/OK/UNSET)

**Comment by ruflin** [comment]
One of the problems `error` always had in ECS was that it was used in too many different ways. Here the current description: 

> Use them for errors that happen while fetching events or in cases where the event itself contains an error.

In addition to the mentioned two use cases above, it is also used for errors happening during ingest. If we already touch error, we should narrow down on what our targeted use case is no matter if it is error or exception. Most cases I see described above are focused on "cases where the event itself contains an error." That is also what I would focus on and scrap "fetching" or "ingest" errors.

**Comment by Oberon00** [comment]
I also think a separate error convention would be the way to go. If we want consistency with spans, it would need to be `status.category` (optional enum with just "OK", "ERROR" and "UNSET" might be represented as actually unset) and `status.message`.

**Comment by pyohannes** [comment]
> I also think a separate error convention would be the way to go.

I understand some of the reasoning behind this, but when I try to picture a concrete solution, it feels quite murky to me:
* Exceptions that are no errors: you'll have `exception.[type,message,stacktrace]`, the span status will be likely be `OK` or `UNSET`, if it wasn't set to `ERROR` due to another reason not related to the recorded exception.
* Errors that are no exceptions: you'll have `error.[code,message,stacktrace]`, and span status will likely be `ERROR`, if the error wasn't handled in the scope of this span or if the error was expected.
* Exceptions that are errors. Should both `exception.*` and `error.*` attributes be set? The span status might be `ERROR` if the exception wasn't handled in the scope of the span, and something else otherwise?

A lot of emphasis is put on exceptions that aren't errors (but instead used for flow control). There might be some controversy about this, but I would argue that recording details about such exceptions has only limited value in the context of telemetry data and should be avoided if possible (what value would we get from e. g. `GeneratorExit` stack traces on spans and logs). With the exception (no pun intended) of Python, this usage of exceptions is also very niche in all languages I'm familiar with.

To summarize, separating errors and exceptions seems cleaner on a conceptual level, however, it adds a fair amount of complexity. The additional value provided seems small to me, especially in the context of telemetry data, and I'm not convinced it is worth the additional complexity.

**Comment by tsloughter** [comment]
T> With the exception (no pun intended) of Python, this usage of exceptions is also very niche in all languages I'm familiar with.

Erlang is another one that uses exceptions for control flow at times and I'd completely agree that these should not be recorded in an `exception.*` attribute. So I support replacing `exception` with `error`.

The only thing I'd add is `error.stacktrace` shouldn't be a required attribute for having an `error` -- in recent Erlang's you can only get the stacktrace if you have an exception :)

**Comment by Oberon00** [comment]
> Errors that are no exceptions: you'll have error.[code,message,stacktrace], and span status will likely be ERROR

For spans, you won't have error.message, since you would store it in the status (which needs to be error then). As for stacktrace, I don't see how it is related to errors in particular. We might want a general stacktrace attribute though (note that the exception's stacktrace attribute is not actually a stacktrace, but the full exception detail representation, usually including exception, causing exception, etc. I protested that name quite vehemently when it was introduced, but since a few languages do call it stacktace, it stuck)

**Comment by trask** [comment]
> What about the Span status? An error can be recorded by setting the span status to ERROR and setting a appropriate status text.

That's a great point that spans already have both concepts:

* `Span.RecordException(Exception)`
* `Span.SetStatus(StatusCode.ERROR, Description)`

So maybe having separate `error.*` attributes would not be so confusing.

As far as introducing `error.*` attributes, how about?

* `error.type` (low-cardinality) - equivalent of Span's StatusCode.ERROR Description
* `error.code` (low-cardinality) - this could be a useful span attribute for systems which have error codes
* `error.message` (not low-cardinality) - this could be useful on spans which want to report a more detailed reason associated with StatusCode.ERROR

(and `error.*` would not need a stack trace attribute, since that's typically the domain of exceptions)

**Comment by Oberon00** [comment]
SGTM, only comment is that maybe "status" would be a better namespace than "error", but I don't have a very strong opinion on that.

**Comment by Oberon00** [comment]
Hmm, I think error.message would be the status description. At least most usages of the status description I have seen don't use a fixed string, but put in some exception message or similar.
Also, I think we need something to represent explicit OK/ERROR status for other telemetry if we want to go this direction. Could be `otel.status`, `error.status`, `status.kind`, etc.

**Comment by trask** [comment]
> Hmm, I think error.message would be the status description. At least most usages of the status description I have seen don't use a fixed string, but put in some exception message or similar.

interesting, we (@open-telemetry/java-instrumentation-maintainers) have interpreted this differently so far: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6710#issuecomment-1448818614

> using exception message contradicts the language used in spec which requires for the description to be predictable and documented. I interpret it to mean that there should be a fixed list of descriptions.

**Comment by trask** [comment]
> I think we need something to represent explicit OK/ERROR status for other telemetry if we want to go this direction.

on the logging side we have [Severity](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#severity-fields), though it doesn't map 1-1 with SpanKind OK/ERROR, and it doesn't solve the issue for metrics signal

**Comment by lmolkova** [comment]
wrt `ote.status_` attributes:

`otel.status_code` definition is very strict and cannot be used as an equivalent of general-purpose `error.code`, but it's probably fine:
- for systems that have status codes, status code is not necessarily an error or success, so they won't use `error.code` for generic code. They can define a system-specific code plus use `otel.status_code` to describe if it's error/success/unknown. 

Functionally, we can extend  `otel.status_code` and `otel.status_description` to be cross-signal attributes, but the whole "otel.*" namespace [seem to be reserved](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#otel-namespace) for internal otel business. Also, given how diverse and mature logging world is, having OTel-specific attributes for errors is a bit too bold.

**Comment by Oberon00** [comment]
@trask Your interpretation may very well be what was originially intended, but it's not generally used that way, e.g. the Python API provides an option to automatically record the exception messge as span status: https://github.com/open-telemetry/opentelemetry-python/blob/v1.17.0/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L951

**Comment by trask** [comment]
@Oberon00 do you think I should raise a spec issue to clarify, or do you think the spec is reasonably clear and I should open a bug report in the python repo? thx

**Comment by Oberon00** [comment]
This is a spec issue IMHO. Python was just one example, the same happens in .NET: https://github.com/open-telemetry/opentelemetry-dotnet/blob/fea6793a5590a18ff6d5e99460f9dfd8cf4c95b2/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs#L248

**Comment by cbandy** [comment]
Some `exception` attributes are "reserved" here: https://github.com/open-telemetry/opentelemetry-specification/blob/7b4c60c6edb52c676d53ced730b6af4ded509068/specification/semantic-conventions.md?plain=1#L10-L23

**Comment by trask** [comment]
`exception.*` attributes are now stable: https://github.com/open-telemetry/semantic-conventions/pull/619
